src/hardware/driver/output.rs

```diff
@@ -0,0 +1,176 @@
+///! Driver output state handling.
+use core::fmt::Debug;
+use embedded_hal::blocking::i2c::{Write, WriteRead};
+use idsp::iir;
+use mcp230xx::{Mcp23008, Mcp230xx};
+use smlang::statemachine;
+
+use super::{relay::Relay, Channel};
+
+pub struct Output<I2C: WriteRead + Write + 'static> {
+    iir: iir::IIR<f32>,
+    relay: Relay<I2C>,
+}
+
+impl<I2C, E> Output<I2C>
+where
+    I2C: WriteRead<Error = E> + Write<Error = E>,
+    E: Debug,
+{
+    // 1 A/s current ramp-up
+    const RAMP_STEP: f32 = 1.0; // will be 1 mA current steps
+    const RAMP_DELAY: fugit::MillisDuration<u64> =
+        fugit::MillisDurationU64::millis(10); // will be current steps every 1 ms
+
+    pub fn new(
+        gpio: &'static spin::Mutex<Mcp230xx<I2C, Mcp23008>>,
+        channel: Channel,
+    ) -> Self {
+        Output {
+            iir: iir::IIR::new(0., -i16::MAX as _, i16::MAX as _), // Todo: sensible defaults for driver (this is for stabilizer now)
+            relay: Relay::new(gpio, channel),
+        }
+    }
+}
+
+pub mod sm {
+    use super::*;
+    statemachine! {
+        transitions: {
+            // Enable sequence
+            *Disabled + Enable / engage_k0 = EnableWaitK0,
+            EnableWaitK0 + RelayDone / disengage_k1 = EnableWaitK1,
+            EnableWaitK1 + RelayDone = RampCurrent,
+            RampCurrent + CurrentStep / increment_current = RampCurrent,
+            RampCurrent + CurrentFinal = Enabled,
+
+            // Abort transitions
+            EnableWaitK0 | EnableWaitK1 | RampCurrent + Disable = Abort,
+            Abort + RelayDone / engage_k1_and_hold_iir = DisableWaitK1,
```

Comment:
What about
`Abort + Disable | Enable`
and
`DisableWaitK1 | DisableWaitK0 + Enable | Disable`
?

Reply:
I was thinking that I'll leave those as errors. It's fairly unlikely in practice that another enable/disable comes within 200 ms after the last (aborting/disabling just takes the two relay transition times). And if this occures it's not a problem since I just print the error that the enable/disable was impossible and nothing bad happens.
https://github.com/quartiq/stabilizer/blob/f0b3e916a8a73f4dcd8d5e9a2bff81c29b2824fe/src/bin/driver.rs#L506-L507